python/mpt/src/mpt/__init__.py

```diff
@@ -0,0 +1,1553 @@
+        nibble_path = nibble_path + b"\x00"  # Pad with zero if odd
+    result = bytes(
+        [
+            nibble_path[i] * 16 + nibble_path[i + 1]
+            for i in range(0, len(nibble_path), 2)
+        ]
+    )
+    return "0x" + result.hex()
+
+
+class StateTries:
+    """
+    Represents the (partial) state of Ethereum for a given block.
+    Includes all account MPT leaf nodes and storage MPT leaf nodes and all intermediary nodes touched in a block,
+    related codes, an access list of addresses and storage keys that were accessed in the block, and a state root.
+    """
+
+    nodes: Mapping[Bytes32, Bytes]
+    codes: Mapping[Bytes32, Bytes]
+    access_list: Mapping[Address, Optional[List[Bytes32]]]
+    state_root: Hash32
+
+    @classmethod
+    def create_empty(cls) -> "StateTries":
+        return cls(
+            nodes={},
+            codes={},
+            access_list={},
+            state_root=EMPTY_TRIE_ROOT_HASH,
+        )
+
+    def __init__(
+        self,
+        nodes: Mapping[Bytes32, Bytes],
+        codes: Mapping[Bytes32, Bytes],
+        access_list: Mapping[Address, Optional[List[Bytes32]]],
+        state_root: Hash32,
+    ):
+        self.nodes = nodes
+        self.codes = codes
+        self.access_list = access_list
+        self.state_root = state_root
+        logger.debug(f"Initialized MPT with state root: 0x{state_root.hex()}")
+        logger.debug(f"Number of nodes: {len(nodes)}")
+        logger.debug(f"Number of codes: {len(codes)}")
+        logger.debug(f"Number of addresses in access list: {len(access_list)}")
+
+    def to_state(self) -> State:
+        """
+        Convert the StateTries to a State object.
+
+        Returns
+        -------
+        State
+            A State object representing the Ethereum state
+        """
+
+        # Step 1: Recursively explore the state trie to get to the leaves
+        _main_trie: Trie[Address, Optional[Account]] = Trie(
+            secured=True, default=None, _data=defaultdict(lambda: None)
+        )
+        _storage_tries: Dict[Address, Trie[Bytes32, U256]] = dict()
+
+        logger.debug("Starting to convert StateTries to State")
+        for address in self.access_list.keys():
+
+            account = self.get(keccak256(address))
+            if account is None:
+                logger.debug(
+                    f"Account not found: {address} - Considering this an exclusion proof"
+                )
+                trie_set(_main_trie, address, EMPTY_ACCOUNT)
+                continue
+
+            rlp_account = RLPAccount(*rlp.decode(account))
+
+            if rlp_account.code_hash == EMPTY_CODE_HASH:
+                code = b""
+            else:
+                code = self.codes.get(rlp_account.code_hash, None)
+                # TODO: This is a hack to get the code for codes not present in the StateTries object
+                # This is due to accessing code only through EXTCODEHASH opcode
```

Comment:
due to the fact that the Account class in python doesn't match the account node structure: the class contains the full code and not only the code hash